### ISSUE TITLE
Glslangvalidator hangs when putting a struct of a struct with multidimensional arrays in a buffer

### ISSUE THREAD

**Issue by doenn** [KhronosGroup/glslang#2842]
I'm running glslangvalidator on a compute shader, which stalls seemingly forever if I have:
1. A buffer
2. Buffer contains a large array of structs
3. Struct contains a large 2 dimensional array

Run cmd:
`.\glslangValidator.exe  -G100 -o .\VKImgBin.spv .\VKImg.comp`

Stripped down source (should be reproducible, I can tweak if needed):

```
#version 450
#extension GL_ARB_separate_shader_objects : enable

layout(set = 0, binding = 1, rgba8) uniform image2D srcImage[1];



struct Buffer1{
	float spherePosI;

};

struct Sphere{
	int boundsIntersect;		// 0 false 1 true

};

struct T2DField{
	vec3 tField[512][256];		// Works when one dimensional [2] or small [2][3]
};

struct S2DField{
	float sField[512][256];
};

struct Media{
	vec3 center;
	vec3 bounds[8];
	T2DField velocity[256];		// This does not work
	T2DField oldVelocity[256];
	S2DField points[256];
	S2DField pressure[256];
	S2DField boundary[256];

};


struct Buffer2{
	float time;
	int firstSetup;
	int sliceIndex;
	Media media;
	//T2DField velocity[256];	// This works, directy in buffer

};





layout(std140, set = 0, binding = 2) readonly buffer B1{
	Buffer1 b1s[];
} b1;

layout(std140, set = 0, binding = 3) buffer B2{
	Buffer2 b2s[];
} b2;




void main() {
	ivec2 pathPos = ivec2(0,0);
	pathPos.x = int(gl_GlobalInvocationID.x);
	pathPos.y = int(gl_GlobalInvocationID.y);

	S2DField pressure[256] = b2.b2s[0].media.pressure;


}
```

**Comment by jeremy-lunarg** [comment]
> which stalls seemingly forever if I have

Are you referring to compile-time or execution-time? I can compile your shader; although, `spirv-val` did produce an error.

**Comment by doenn** [comment]
OK, maybe it had to be used. I updated above with this line:

`S2DField pressure[256] = b2.b2s[0].media.pressure;`

in the main(). I just tried it now twice and it hangs. 

Some other notes, somewhat unrelated:

- I am not blocked by this anymore, I can compile with newer target-env:

`.\glslangValidator.exe  -G100 --target-env vulkan1.2 -o .\VKImgBin.spv .\VKImg.comp`

And had to eventually put these large multi-dimensional arrays in separate StorageBuffers, i.e.:

```
layout(std430, set = 0, binding = 2) buffer B1{
	Buffer1 b1s[];
} b1;

layout(std430, set = 0, binding = 3) buffer B2{
	Buffer2 b2s[];
} b2;

layout(std430, set = 0, binding = 4) buffer B3{
	Buffer3 b3s[];
} b3;
```

- Render Doc showed an error that some members were overwriting others or something, so I changed to std430 to make the layout less 'relaxed'. 
- I can't pass these multi-dimensional arrays to functions I think because it's making them temporary storage and they are too large, cant remember if their size in functions was a runtime issue but I think so. Maybe there's a way to pass ptrs instead. If so, I will figure it out.
- I'm also seeing: `Core     PID  14748: [00:26:46]    spirv_reflect.cpp(1426) - Error   - Assertion failed: 'decorations[curType->id].arrayStride <= 0xffff' Stride is too large for uint16_t` Not sure where this comes from, but I've broken the struct down into a single dimension struct that contains a two dimensional. If array indices are **uint16**, maybe this will solve since perhaps it can use the first dimension index, then following two when indexing, not have to use dim1xdim2xdim3 which > max **uint16**. I can read SPIRV spec to see if this is how indexing chains works or not.
- When targeting `--target-env vulkan1.2` I now no longer can pass uniform images to functions as writeonly. Had to inline these functions which contain `imageStore()` in main(). Minor inconvenience but step backwards. I'm aware there are separate issues open for this.

I'm getting stuff running now, though everything is kinda messy inlined in the main(), but it's at least working. Just want to share these issues I ran into.

**Comment by jeremy-lunarg** [comment]
This shader isn't hanging; however, it takes a very long time to compile if targeting `spir-v 1.3` or below. The target environment `vulkan1.1` defaults to `spir-v 1.3`, which means that copying the array will generate instructions (i.e. `OpCompositeExtract`, `OpAccessChain`, and `OpStore`) for _each_ element of the array. Using the target environment `vulkan1.2` defaults to `spir-v 1.5`, which can generate a single `OpCopyLogical` instruction instead.

FWIW, you can ask `glslangValidator` to generate `spir-v 1.4` or `spir-v 1.5` under target environment `vulkan1.0` or `vulkan1.1`. For example:

`./glslangValidator --target-env vulkan1.0 --target-env spirv1.4 test_issue_2842.comp -o test_issue_2842.comp`

**Comment by jeremy-lunarg** [comment]
>     * I'm also seeing: `Core     PID  14748: [00:26:46]    spirv_reflect.cpp(1426) - Error   - Assertion failed: 'decorations[curType->id].arrayStride <= 0xffff' Stride is too large for uint16_t` Not sure where this comes from, but I've broken the struct down into a single dimension struct that contains a two dimensional. If array indices are **uint16**, maybe this will solve since perhaps it can use the first dimension index, then following two when indexing, not have to use dim1xdim2xdim3 which > max **uint16**. I can read SPIRV spec to see if this is how indexing chains works or not.

Can you file this issue in the [SPIRV-Reflect](https://github.com/KhronosGroup/SPIRV-Reflect) repository?

**Comment by jeremy-lunarg** [comment]
> When targeting `--target-env vulkan1.2` I now no longer can pass uniform images to functions as writeonly. Had to inline these functions which contain `imageStore()` in main(). Minor inconvenience but step backwards. I'm aware there are separate issues open for this.

Can you create a new issue for this with an example shader? Thanks!
